File: TypeScript/src/session2.td/agregation/EtatNaturelPur.ts

import {FabriqueNaturels} from '../../session1.td/FabriqueNaturels';
import {EtatNaturel} from './EtatNaturel';
import {ZeroInductif} from './ZeroInductif';

export interface EtatNaturelPur extends EtatNaturel<EtatNaturelPur>, FabriqueNaturels<EtatNaturelPur> {

    creerNatAvecValeur(val : number) : EtatNaturelPur;
    creerZero() : EtatNaturelPur;
    creerSuccesseur(predecesseur : EtatNaturelPur ) : EtatNaturelPur;
    creerNatAvecRepresentation(repDecimale : string) : EtatNaturelPur;
}

export abstract class EtatNaturelPur_default implements EtatNaturelPur {

    creerNatAvecValeur(val : number) : EtatNaturelPur {
        return this.creerNatAvecRepresentation(String(val));
    }

    creerZero() : EtatNaturelPur {
        return this.creerNatAvecValeur(0);
    }

    creerSuccesseur(predecesseur : EtatNaturelPur ) : EtatNaturelPur {
        return this.creerNatAvecValeur(predecesseur.val());
    }

    creerNatAvecRepresentation(repDecimale : string) : EtatNaturelPur {
        return this.creerNatAvecValeur(+repDecimale);
    }

    abstract chiffre(n: number): number;
    abstract estNul(): boolean;
    abstract predecesseur(): EtatNaturelPur;
    abstract taille(): number;
    abstract val(): number;
}

File: Java/src/session2/td/agregation/EtatNaturelPur.java
package session2.td.agregation;

import session1.td.FabriqueNaturels;

public interface EtatNaturelPur extends EtatNaturel<EtatNaturelPur>, FabriqueNaturels<EtatNaturelPur> {

    @Override
    default EtatNaturelPur creerNatAvecValeur(int val) {
        if(val == 0)
            return new ZeroInductif();
        return this.creerSuccesseur(this.creerNatAvecValeur(val -1));
    }
    default EtatNaturelPur creerZero() {
        return this.creerNatAvecValeur(0);
    }
    default EtatNaturelPur creerSuccesseur(EtatNaturelPur predecesseur) {
        return new IntPositif(predecesseur.val() + 1);
    }
    default EtatNaturelPur creerNatAvecRepresentation(String repDecimale) {
        return this.creerNatAvecValeur(Integer.parseInt(repDecimale));
    }
}
